9856 Replace graphene with Strawberry

arthanson opened the pull request:
### Fixes: #9856 

Replaces django-graphene with strawberry-django a new engine for handling GraphQL queries.  Notable changes:

* Converts all GraphQL types to Strawberry Types
* Includes much more updated GraphiQL browser
* Decorator to auto-generate Strawberry required data-classes for filters as Strawberry doesn't support django-filter currently.

Example Queries:
```
{
  circuit_termination_list(filters: {cabled: false}) {
    id
    display
    cable {
      id
    }
  }
}
```
and
```
{
  location_list {
    image_attachments {
      id
    }
    contacts {
      id
    }
    changelog {
      id
    }
    display
    class_type
    tags {
      id
    }
    custom_fields
    id
    created
    last_updated
    custom_field_data
    name
    slug
    description
    site {
      id
    }
    status
    tenant {
      id
    }
    lft
    rght
    tree_id
    level
    children {
      id
    }
    powerpanel_set {
      id
    }
    cabletermination_set {
      id
    }
    racks {
      id
    }
    vlan_groups {
      id
    }
    parent {
      id
    }
    devices {
      id
    }
  }
}
```
### Filtering lookups are different between the two libraries:

**Graphene**
```
user_list(is_superuser: true) {
    id
}
```
**Strawberry**
```
user_list(filters: {is_superuser: true}) {
   id
}
```
The filtering lookup syntax (greater than, less than, case-insensitive compare) is also different:
**Graphene**
```
{
  user_list(username__ic: "bob") {
    username
  }
}
```
**Strawberry**
```
{
  user_list(filters: {username: {i_contains: "bob"}}) {
    username
  }
}
```

For stawberry, allowing lookups on fields forces the lookup to be present, so it forces this somewhat awkward syntax:
**Graphene**
```
{
  user_list(id: "3") {
    id
  }
}
```
**Strawberry**
```
{
  user_list(filters: {id: {exact: 3}}) {
    id
  }
}
```
## Performance optimizer

Below is a ridiculous nested query that can be used to show optimization

**old graphene:** 55 queries
**strawberry:** 24 queries

```
{
  location_list {
    image_attachments {
      id
    }
    contacts {
      id
    }
    changelog {
      id
    }
    display
    class_type
    tags {
      id
    }
    custom_fields
    id
    created
    last_updated
    custom_field_data
    name
    slug
    description
    site {
      id
      name
      locations {
        id
        name
      }
    }
    status
    tenant {
      id
      sites {
        id
        name
      }
    }
    lft
    rght
    tree_id
    level
    children {
      id
    }
    powerpanel_set {
      id
    }
    cabletermination_set {
      id
    }
    racks {
      id
      site {
        id
        name
      }
    }
    vlan_groups {
      id
    }
    parent {
      id
    }
    devices {
      id
    }
  }
}
```

arthanson committed:
9856 base strawberry integration

arthanson committed:
9856 user and group

arthanson committed:
9856 user and circuits base

arthanson committed:
9856 extras and mixins

arthanson committed:
9856 fk

arthanson committed:
9856 merge feature

arthanson committed:
Merge branch 'feature' into 9856-strawberry-2

arthanson committed:
9856 update strawberry version

arthanson committed:
9856 update imports

arthanson committed:
9856 compatability fixes

arthanson committed:
9856 compatability fixes

arthanson committed:
9856 merge feature

arthanson committed:
9856 update strawberry types

arthanson committed:
9856 update strawberry types

arthanson committed:
9856 core schema

arthanson committed:
9856 dcim schema

arthanson committed:
9856 extras schema

arthanson committed:
9856 ipam and tenant schema

arthanson committed:
9856 virtualization, vpn, wireless schema

arthanson committed:
9856 fix old decorator

arthanson committed:
Merge branch 'feature' into 9856-strawberry-2

arthanson committed:
9856 cleanup

arthanson committed:
9856 cleanup

arthanson committed:
9856 fixes to circuits type specifiers

arthanson commented:
Remaining work:

- [ ] Validate Circuits Types
- [ ] Validate Circuits Fitlersets
- [ ] Auto generate Filtersets definitions
- [ ] Choice field handling
- [ ] UsersQuery
- [ ] CoreQuery
- [ ] DCIMQuery
- [ ] ExtrasQuery
- [ ] IPAMQuery
- [ ] TenancyQuery
- [ ] VirtualizationQuery
- [ ] VPNQuery
- [ ] WirelessQuery
- [ ] Fix GraphQL tests auto-discovery of fields
- [ ] Update netbox/graphql/views.py GraphQLView
- [ ] netbox/netbox/graphql/fields.py (ObjectField, ObjectListfield)? - I think this is obsolete from optimizer
- [ ] netbox/netbox/graphql/scalars.py (BigInt)?
- [ ] netbox/netbox/graphql/utils.py (get_graphene_type)?
- [ ] netbox/dcim/graphql/gfk_mixins.py (gfk mixins)?

arthanson committed:
9856 fixes to circuits type specifiers

arthanson committed:
Merge branch 'feature' into 9856-strawberry-2

arthanson committed:
9856 update types

arthanson commented:
Including some GraphQL queries here I used for testing as they are handy for reference - just got some GFK stuff working! :tada: :

```
  {
    circuit_termination_list {
      id
      link_peers {
        __typename
        ... on InterfaceType {
          id
          name
        }
        ... on RearPortType {
          id
          display
          device {
            pk
          }
        }
      }
    }
  }
```

arthanson committed:
9856 GFK working

arthanson committed:
9856 GFK working

arthanson committed:
9856 _name

arthanson committed:
9856 misc fixes

arthanson committed:
Merge branch 'feature' into 9856-strawberry-2

arthanson committed:
9856 type updates

arthanson committed:
9856 _name to types

arthanson committed:
9856 update types

arthanson committed:
9856 update types

arthanson committed:
9856 update types

arthanson committed:
9856 update types

arthanson committed:
9856 update types

arthanson committed:
9856 update types

arthanson committed:
9856 update types

arthanson committed:
9856 update types

arthanson committed:
9856 update types

arthanson committed:
9856 merge feature

arthanson committed:
9856 GraphQLView

arthanson committed:
9856 GraphQLView

arthanson committed:
Merge branch 'feature' into 9856-strawberry-2

arthanson committed:
9856 fix OrganizationalObjectType

arthanson committed:
9856 single item query for schema

arthanson committed:
9856 circuits graphql tests working

arthanson committed:
9856 test fixes

arthanson committed:
9856 test fixes

arthanson committed:
9856 test fixes

arthanson committed:
9856 test fix vpn

arthanson committed:
9856 test fixes

arthanson committed:
9856 test fixes

arthanson committed:
9856 test fixes

arthanson committed:
9856 circuits test sans DjangoModelType

arthanson committed:
9856 core test sans DjangoModelType

arthanson committed:
9856 temp checkin

arthanson committed:
9856 fix extas FK

arthanson committed:
9856 fix tenancy FK

arthanson committed:
9856 fix virtualization FK

arthanson committed:
9856 fix vpn FK

arthanson committed:
9856 fix wireless FK

arthanson committed:
9856 fix ipam FK

arthanson committed:
9856 fix partial dcim FK

arthanson committed:
9856 fix dcim FK

arthanson committed:
9856 fix virtualization FK

arthanson committed:
9856 fix tests / remove debug code

arthanson committed:
9856 fix test imagefield

arthanson committed:
9856 cleanup graphene

arthanson committed:
9856 fix plugin schema

arthanson committed:
9856 fix requirements

arthanson committed:
9856 fix requirements

arthanson committed:
9856 fix docs

arthanson committed:
9856 fix docs

arthanson committed:
9856 temp fix tests

arthanson committed:
9856 first filterset

arthanson committed:
9856 first filterset

arthanson committed:
9856 merge feature

arthanson committed:
9856 fix tests

arthanson committed:
9856 fix tests

arthanson committed:
9856 working auto filter generation

arthanson committed:
9856 filter types

arthanson committed:
9856 filter types

arthanson committed:
9856 filter types

arthanson committed:
9856 fix graphiql test

arthanson committed:
Merge branch 'feature' into 9856-strawberry-2

arthanson committed:
9856 fix counter fields and merge feature

arthanson committed:
9856 temp fix tests

arthanson committed:
9856 fix tests

arthanson committed:
9856 fix tenancy, ipam filter definitions

arthanson committed:
9856 cleanup

arthanson committed:
9856 cleanup

arthanson committed:
9856 cleanup

jeremystretch committed:
Merge branch 'feature' into 9856-strawberry-2

jeremystretch commented on netbox/utilities/testing/api.py:
```diff
@@ -447,34 +450,34 @@ def _build_query(self, name, **filters):
-                    # Union types dont' have an id or consistent values
+
+            for field in type_class.__strawberry_definition__.fields:
+                if (
+                    field.type in (
+                        strawberry_django.fields.types.DjangoFileType, strawberry_django.fields.types.DjangoImageType) or
+                    type(field.type) is StrawberryOptional and field.type.of_type in (
+                        strawberry_django.fields.types.DjangoFileType, strawberry_django.fields.types.DjangoImageType)
+                ):
```
We can clean this up by defining a list of fields to check for.
```suggestion
                file_fields = (strawberry_django.fields.types.DjangoFileType, strawberry_django.fields.types.DjangoImageType)
                if (
                    field.type in file_fields or (
                        type(field.type) is StrawberryOptional and field.type.of_type in file_fields
                    )
                ):
```

jeremystretch commented on netbox/core/graphql/types.py:
```diff
@@ -1,21 +1,34 @@
 
+@strawberry_django.type(
+    models.DataFile,
+    exclude=('data',),
```
`strawberry_django` seems to demand a list here.
```suggestion
    exclude=['data'],
```

jeremystretch commented on netbox/circuits/graphql/filters.py:
```diff
@@ -0,0 +1,50 @@
+
+from netbox.graphql.filter_mixins import autotype_decorator, BaseFilterMixin
+
+__all__ = (
```
Should we look into automating the generation of these filters? Or would you prefer to leave it explicit for now to avoid potentially complicating things?

arthanson replied:
I think leave it explicit for now as you may want to possibly add something to the filter beyond the auto-generated response.

jeremystretch commented on netbox/dcim/graphql/types.py:
```diff
@@ -67,447 +79,944 @@ class ComponentObjectType(
-        return self.length_unit or None
+@strawberry_django.type(
+    models.CableTermination,
+    exclude=('termination_type', 'termination_id'),
```
`exclude` needs to be a list to make type checking happy.
```suggestion
    exclude=['termination_type', 'termination_id'],
```

jeremystretch commented on netbox/dcim/graphql/types.py:
```diff
@@ -67,447 +79,944 @@ class ComponentObjectType(
 
+@strawberry_django.type(
+    models.Location,
+    # fields='__all__',
```
```suggestion
```

jeremystretch commented on netbox/dcim/graphql/types.py:
```diff
@@ -67,447 +79,944 @@ class ComponentObjectType(
+@strawberry_django.type(
+    models.Region,
+    exclude=('parent',),
+    # fields='__all__',
```
```suggestion
```

jeremystretch commented on netbox/dcim/graphql/types.py:
```diff
@@ -67,447 +79,944 @@ class ComponentObjectType(
+
+@strawberry_django.type(
+    models.SiteGroup,
+    # fields='__all__',
```
```suggestion
```

jeremystretch commented on netbox/netbox/graphql/filter_mixins.py:
```diff
@@ -0,0 +1,194 @@
+        if fieldname not in cls.__annotations__ and attr_type:
+            cls.__annotations__[fieldname] = attr_type
+
+        fname = f"filter_{fieldname}"
```
Rename `fname` to `filter_name` for clarity?

jeremystretch commented on netbox/netbox/graphql/filter_mixins.py:
```diff
@@ -0,0 +1,194 @@
+            attr_type = None
+
+            # NetBox Filter types - put base classes after derived classes
+            if isinstance(field, ContentTypeFilter):
```
Might be cleaner to move this swath of logic into a separate function.

jeremystretch commented on netbox/tenancy/graphql/types.py:
```diff
@@ -14,64 +18,177 @@
 
 
+@strawberry.type
 class ContactAssignmentsMixin:
```
Should we move this to `tenancy/mixins.py` for consistency with other apps?

jeremystretch commented on netbox/users/graphql/types.py:
```diff
@@ -1,38 +1,45 @@
-    def get_queryset(cls, queryset, info):
-        return RestrictedQuerySet(model=Group).restrict(info.context.user, 'view')
+    def get_queryset(cls, queryset, info, **kwargs):
+        return RestrictedQuerySet(model=Group).restrict(info.context.request.user, 'view')
```
Is a custom `get_queryset()` method still needed for the User & Group models now that we have our own custom models in place?

jeremystretch commented on requirements.txt:
```diff
@@ -31,6 +30,7 @@ PyYAML==6.0.1
 requests==2.31.0
 social-auth-app-django==5.4.0
 social-auth-core[openidconnect]==4.5.3
+strawberry-graphql-django==0.33.0
```
We should add `strawberry` itself too as a direct dependency.

jeremystretch reviewed (changes requested):
Excellent work! Noticed a few minor cleanup items on initial review and had some general questions.

arthanson commented on netbox/circuits/graphql/types.py:
```diff
@@ -15,48 +20,93 @@
-        model = models.CircuitTermination
-        fields = '__all__'
-        filterset_class = filtersets.CircuitTerminationFilterSet
+    @strawberry_django.field
+    def networks(self) -> List[Annotated["ProviderNetworkType", strawberry.lazy('circuits.graphql.types')]]:
+        return self.networks.all()
```
Its going to ProviderNetworkType, this is ProviderType

jeremystretch replied:
What I mean is `ProviderType.networks` seems to return itself, for instance, because of the decorator. It's not a problem, just a bit confusing at first glance.

arthanson committed:
9856 review changes

arthanson committed:
9856 review changes

arthanson committed:
9856 review changes

arthanson committed:
9856 fix base-requirements

arthanson commented on base_requirements.txt:
```diff
@@ -136,8 +131,16 @@ social-auth-core
+
+# Strawberry GraphQL Django extension
+# https://github.com/strawberry-graphql/strawberry-django/blob/main/CHANGELOG.md
+strawberry-django
```
oops, yes

jeremystretch reviewed (changes requested):
The GraphiQL view isn't using our custom template, and static resources are being pulled from unpkg.com rather than being served locally.

Also, it looks like maybe we should swap `django-graphiql-debug-toolbar` for `django-graphiql-strawberry-debug-toolbar`, although AFAICT the current middleware seems functional.

arthanson committed:
9856 add wrapper to graphiql

arthanson committed:
9856 remove old graphiql debug toolbar

arthanson commented:
Fixed up the GraphiQL view, the old method of bundling the assets wouldn't work as the new GraphiQL has some include packages that ESBuild can't translate to ES6, see: (https://github.com/graphql/graphiql/discussions/2525) - we could switch to webpack which does handle it, but as it is just for GraphiQL it seemed fine to just statically link the separate files.  Changed the ESBuild script to copy the files out of the node-packages.

django-graphical-debug-toolbar is no longer needed, so removed that.  The strawberry one is handled by middleware already.

jeremystretch commented on netbox/netbox/settings.py:
```diff
@@ -474,6 +473,10 @@ def _setting(name, default=None):
     os.path.join(BASE_DIR, 'project-static', 'img'),
     os.path.join(BASE_DIR, 'project-static', 'js'),
     ('docs', os.path.join(BASE_DIR, 'project-static', 'docs')),  # Prefix with /docs
+    # os.path.join(ROOT_DIR, 'node_modules', 'graphiql-explorer'),
+    # os.path.join(ROOT_DIR, 'node_modules', 'react', 'cjs'),
+    # os.path.join(ROOT_DIR, 'node_modules', 'react_dom', 'cjs'),
+    # os.path.join(ROOT_DIR, 'node_modules', 'js-cookie', 'dist'),
```
Safe to remove these?
```suggestion
    # os.path.join(ROOT_DIR, 'node_modules', 'graphiql-explorer'),
    # os.path.join(ROOT_DIR, 'node_modules', 'react', 'cjs'),
    # os.path.join(ROOT_DIR, 'node_modules', 'react_dom', 'cjs'),
    # os.path.join(ROOT_DIR, 'node_modules', 'js-cookie', 'dist'),
```

jeremystretch commented on netbox/project-static/bundle.js:
```diff
@@ -14,24 +17,49 @@ const options = {
+  fileMap = [
+    {
+      source: './node_modules/react/umd/react.production.min.js',
+      dest: './dist/react.production.min.js'
```
Can we move these to a subdirectory ( e.g. `static/graphiql/`) to keep them isolated from "native" static resources?

jeremystretch commented on netbox/project-static/bundle.js:
```diff
@@ -14,24 +17,49 @@ const options = {
-  } catch (err) {
-    console.error(err);
-  }
+  ]
```
```suggestion
  ];
```

jeremystretch commented on netbox/project-static/bundle.js:
```diff
@@ -14,24 +17,49 @@ const options = {
-    });
-    if (result.errors.length === 0) {
-      console.log(`✅ Bundled source file 'netbox-graphiql/index.ts' to 'graphiql.js'`);
+  fileMap = [
```
```suggestion
  let fileMap = [
```

jeremystretch reviewed (changes requested):
Almost there!

arthanson committed:
Merge branch 'feature' into 9856-strawberry-2

arthanson committed:
9856 review changes

arthanson committed:
9856 update strawberry

arthanson committed:
9856 remove superfluous check
